Inspect updates, including new article for C# (#157)
ShiftLeftSecurity/documentation: using-inspect-protect/inspect/analyzing-applications.md

If you are using ShiftLeft Inspect with ShiftLeft Protect, you can analyze your code before running your application, for example to integrate with a build / Continuous Integration (CI) environment. To do so, you must integrate the [ShiftLeft JSON file](../protect/json-file.md) generated by ShiftLeft Inspect and including parameters required by the ShiftLeft Protect Microagent to customize the monitoring and securing of your application, into your runtime environment and make it available to ShiftLeft Protect. The ShiftLeft JSON file by default is located in the directory from which you run the analysis.

The ShiftLeft Inspect analysis commands are different depending on whether your application is written in:

* [Java and Scala](analyze-java.md)
* [C#](analyze-csharp.md)

## Analysis Workflow

1. Review the [ShiftLeft Inspect requirements](../../introduction/requirements.md).
2. [Install the ShiftLeft Command Line Interface (CLI)](../using-cli/install-cli.md) on the host from where you submit applications for analysis.

> 3. Successfully build the application using a supported build tool (maven, gradle, sbt) **before** you submit the application to ShiftLeft Inspect for analysis. 

Reviewer: This is only necessary for Java. For C#, it's not required that the application has been built.

Author: Added

4. Submit the application for analysis using the appropriate analysis command and options, including [CPG mode](#cpg-mode).
5. For each code commit, rebuild the application and resubmit it for analysis. You can automate this process by integrating the ShiftLeft CLI with a CI/CD build system, for example with [Jenkins](../integrating-with-shiftleft/integrating-jenkins-builds/integrating-jenkins-builds.md), [GoCD](../integrating-with-shiftleft/integrating-gocd-builds.md), [Bamboo](../integrating-with-shiftleft/integrating-bamboo-builds.md), [TeamCity](../integrating-with-shiftleft/integrating-teamcity-builds.md), [Travis](../integrating-with-shiftleft/integrating-travis-builds.md) and CircleCI.

## CPG Mode

> The [CPG mode](options/analyze-options.md) can be used optionally for [analyzing Java and Scala applications](analyze-java.md), and is required for [analyzing C# applications](analyze-csharp.md). This mode builds the CPG locally (on your machine) and then uploads the CPG, rather than your application's code, to the ShiftLeft cloud for analysis. This approach is useful for proprietary application code (in lieu of bytecode), and bytecode for open source dependencies. 

Reviewer: ```suggestion
The [CPG mode](options/analyze-options.md) can be used optionally for [analyzing Java and Scala applications](analyze-java.md), and is automatically implied for [analyzing C# applications](analyze-csharp.md). This mode builds the CPG locally (on your machine) and then uploads the CPG, rather than your application's code, to the ShiftLeft cloud for analysis.
```

Author: Done

Include the option 

```
--cpg
```

> in the `sl analyze` command. For example `sl analyze --app <name> --cpg --csharp <target platform> --csharp2cpg-args “-l trace --trace-all” [<path>]`.

Reviewer: Best to give a java example, as for .NET, it's anyway implied.

Author: Can you provide me with an appropriate Java example?

Reviewer: Or actually, this example can be generic:

`sl analyze --cpg ...`

Author: Done

> **Note**: The CPG mode consumes more memory on your build machine. Ensure that you have at least 16GB RAM available if you plan to use this mode.

Reviewer: This is no longer the case - we can remove this Note altogether.

Author: Removed
